math/src/polynom/multi_linear.rs

```diff
@@ -0,0 +1,244 @@
+///
+/// [`EqFunction`] represents EQ^{~} the multi-linear extension of
+///
+/// $((y_0, ..., y_{ν - 1}) ↦ EQ((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1})))$
+///
+/// and contains a method to generate the Lagrange kernel for defining evaluations of multi-linear
+/// extensions of arbitrary functions $(f: {0 , 1}^ν ⇾ 𝔽)$ at a given point $(r_0, ..., r_{ν - 1})$
+/// as well as a method to evaluate $EQ^{~}((r_0, ..., r_{ν - 1}), (t_0, ..., t_{ν - 1})))$ for
+/// $(t_0, ..., t_{ν - 1}) ∈ 𝔽^ν$.
+pub struct EqFunction<E> {
+    r: Vec<E>,
+}
+
+impl<E: FieldElement> EqFunction<E> {
+    /// Creates a new [EqFunction].
+    pub fn new(r: Vec<E>) -> Self {
+        let tmp = r.clone();
+        EqFunction { r: tmp }
+    }
+
+    /// Computes $EQ((r_0, ..., r_{ν - 1}), (t_0, ..., t_{ν - 1})))$.
+    pub fn evaluate(&self, t: &[E]) -> E {
+        assert_eq!(self.r.len(), t.len());
+
+        (0..self.r.len())
+            .map(|i| self.r[i] * t[i] + (E::ONE - self.r[i]) * (E::ONE - t[i]))
+            .fold(E::ONE, |acc, term| acc * term)
+    }
+
+    /// Computes $EQ((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1}))$ for all
+    /// $(y_0, ..., y_{ν - 1}) ∈ {0 , 1}^ν$ i.e., the Lagrange kernel at $r = (r_0, ..., r_{ν - 1})$.
+    pub fn evaluations(&self) -> Vec<E> {
+        compute_lagrange_basis_evals_at(&self.r)
+    }
+
+    /// Returns the evaluations of
+    /// $((y_0, ..., y_{ν - 1}) ↦ EQ^{~}((r_0, ..., r_{ν - 1}), (y_0, ..., y_{ν - 1})))$
+    /// over ${0 , 1}^ν$.
+    pub fn ml_at(evaluation_point: Vec<E>) -> MultiLinearPoly<E> {
+        let eq_evals = EqFunction::new(evaluation_point.clone()).evaluations();
+        MultiLinearPoly::from_evaluations(eq_evals)
+    }
+}
+
+// HELPER
+// ================================================================================================
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set ${0 , 1}^ν$ at $(r_0, ..., r_{ν - 1})$ i.e., the Lagrange kernel at $(r_0, ..., r_{ν - 1})$.
+fn compute_lagrange_basis_evals_at<E: FieldElement>(query: &[E]) -> Vec<E> {
+    let nu = query.len();
+    let n = 1 << nu;
+
+    let mut evals: Vec<E> = vec![E::ONE; n];
+    let mut size = 1;
+    for r_i in query.iter().rev() {
+        size *= 2;
+        for i in (0..size).rev().step_by(2) {
+            let scalar = evals[i / 2];
+            evals[i] = scalar * *r_i;
+            evals[i - 1] = scalar - evals[i];
+        }
+    }
+    evals
+}
+
+/// Computes the inner product in the extension field of two iterators that must yield the same
+/// number of items.
+pub fn inner_product<E: FieldElement>(x: &[E], y: &[E]) -> E {
+    #[cfg(not(feature = "concurrent"))]
+    return x.iter().zip(y.iter()).fold(E::ZERO, |acc, (x_i, y_i)| acc + *x_i * *y_i);
+
+    #[cfg(feature = "concurrent")]
+    return x
+        .par_iter()
+        .zip(y.par_iter())
+        .map(|(x_i, y_i)| *x_i * *y_i)
+        .reduce(|| E::ZERO, |a, b| a + b);
+}
```

Comment:
Nice!

I would mention in the doc comments that this function is uses multiple threads when `concurrent` feature is enabled.